### PR TITLE
A dev guide

### LINKED ISSUE #68: Print Template
As a CEG/EE Student, I want to **print the current template** I am working on so that I can Find out my current configuration.

### PR DESCRIPTION
**Pull request by sevenseasofbri**
Refactor code
Move gates out of component
Move BinaryTree to new package under model
Update Dev Guide
Add BinaryTree implementation for Boolean Logic Circuits
Details on methods accessed by BooleanTemplate
Might have to check the Sequence Diagrams

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -45,6 +45,81 @@ The `Model`,
...
+## Implementation of Boolean Commands
+
+### Binary Tree
+The boolean add, set and calculate features are modeled using a generic `BinaryTree<T>` class. The `BooleanTemplate` imports
+this class to store and evaluate the logic circuit. 
+
+The elements of the tree are stored in a fixed `ArrayList` (size = 15) indexed in a _heap-like_
+manner. That is, a left to right _level-order traversal_ will map to the indexes of the array. The following diagram represents
+the indexes each node in the tree corresponds to in the `ArrayList`.
+```
```
Comment (hughjazzman) [inline docs/DeveloperGuide.md:80]: Nice diagram!

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -45,6 +45,81 @@ The `Model`,
...
+* `BinaryTree#isLeaf(int)` - Returns boolean based on whether node at specified index is a _leaf node_ or not.
+* `BinaryTree#isEmpty()` - Checks if the tree has no elements in it.
+
+When an object of the `BinaryTree<T>` class is created, it initialises the ArrayList<T> instance to 15 `null` values. This will be 
+further discussed in the section detailing the `insert()` function.
+
+#### Initialising A `BinaryTree<T>` Object
+The Logic initialises the `BinaryTree<Gate>` object using the parameterised constructor, thus specifying the Gate class type 
+root. The object diagram below depicts the initial state of the Model when a `BinaryTree<Gate>` object is created.
+![InitialBinaryTree](diagrams/BinaryTreeInitialObjectDiagram.png)
```
Comment (hughjazzman) [inline docs/DeveloperGuide.md:103]:
```suggestion

![InitialBinaryTree](diagrams/BinaryTreeInitialObjectDiagram.png)
```

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -68,10 +68,83 @@ The `Model`,
...
+
+The Logic uses the parameterised constructor of `BinaryTree<T>` to create the object since it requires initialisation of
+the root. Such an object is created as follows: `BinaryTree<Gate> obj = new BinaryTree(new OrGate(1,1))`. This sets the root
+of the Binary Tree to the object specified.
+
+#### Using `BinaryTree#isNullAtIndex(int)`
+`BooleanTemplate` uses this function to render the current configuration of the circuit in a String format. Besides, the method
+is extensively used in other internal operations in `BinaryTree<T>` for checking whether a position in the tree has been set or not.
```
Comment (hughjazzman) [inline docs/DeveloperGuide.md]:
```suggestion
`BooleanTemplate` uses this function to render the current configuration of the circuit in a String format. The method
is also extensively used in other internal operations in `BinaryTree<T>` for checking whether a position in the tree has been set or not.
```

**Review by hughjazzman — APPROVED** [review]
Sexy UML diagrams.

**Comment by sevenseasofbri** [comment]
> Sexy UML diagrams.

LMAO DHSFBJEFKERF

**Comment by sevenseasofbri** [comment]
BTW I'll add the glossary after everyone has added their parts to the dev guide!

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -68,10 +68,84 @@ The `Model`,
 * has `Component` and `Gate` within the templates.
 * does not depend on any of the other three components.
 
-## Appendix: Requirements
```
Comment (hughjazzman) [inline docs/DeveloperGuide.md:71]: Just realised this shouldn't have been deleted.

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -68,10 +68,84 @@ The `Model`,
...
+![IsEmptySequenceDiagram](diagrams/BinaryTreeIsEmptySequenceDiagram.png)
+ 
+### Rendering Current Boolean Circuit State
+Using a _standard I/O operation_ (Like _Sopln()_) on an object of the `BooleanTemplate` class yields the current configuration
+of the system.
+Each node of the system is represented by a signal ranging from B to O (OUT being the root). All nodes with null parent nodes are
+not shown in the diagram.
+
+## Product scope
+### Target user profile
```
Comment (hughjazzman) [inline docs/DeveloperGuide.md:148]: Should be
```suggestion
## Appendix: Requirements
### Product scope
#### Target user profile
```

Comment (sevenseasofbri) [reply docs/DeveloperGuide.md:148]: Oh no oops, perhaps I'll add it in a next PR?

**Pull request merged**